[#43] 공통 API 응답 객체 생성

wooni97 opened the pull request.

wooni97 committed:
feat: 공통 API 응답 객체 생성

wooni97 committed:
feat: ListApiResponse 메서드 구현

wooni97 committed:
refactor: 공통 응답 객체 구현에 따른 controller response 수정

wooni97 committed:
feature: MemberResponse 클래스 생성

wooni97 committed:
feature: 스터디룸 생성 요청 객체 생성

wooni97 committed:
feature: 스터디룸 api response 객체 생성

wooni97 committed:
refactor: 공통 API 응답 객체 생성에 따른 controller response 수정

thxwelchs commented on src/main/java/dev/flab/studytogether/domain/member/controller/MemberAPIController.java:
```diff
@@ -1,22 +1,18 @@
 
-
-@RequestMapping("/members")
+@RequestMapping("/api/members")
```
/api prefix를 붙인 이유가 무엇일까요?

thxwelchs commented on src/main/java/dev/flab/studytogether/domain/member/controller/MemberAPIController.java:
```diff
@@ -30,32 +26,39 @@ public MemberAPIController(MemberService memberService) {
+        Member member = memberService.createMember(requestDto.getId(), requestDto.getPassword(), requestDto.getNickname());
+        MemberResponse response = new MemberResponse(member.getSequenceId(), member.getId(), member.getNickname());
+
+        return SingleApiResponse.ok(response);
```
기존 구조에서는 login 페이지로 리다이렉트 하는 view 역할이었는데, 이 부분이 API 구조로 변경되는것을 의도하신걸까요?

thxwelchs commented on src/main/java/dev/flab/studytogether/domain/member/dto/MemberResponse.java:
```diff
@@ -0,0 +1,10 @@
+
+import lombok.Data;
+
+@Data
```
@Data 어노테이션은 어떤 역할을 하는것일까요? 이것을 사용함으로써 불필요한 어노테이션이 같이 사용되고 있진 않을까요?

thxwelchs commented on src/main/java/dev/flab/studytogether/response/ListApiResponse.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.List;
+
+public class ListApiResponse<T> extends CommonResponse {
```
SingleApiResponse와 비교해보았을 때, message 필드 보유 여부 외에는 큰 차이가 없어보이는데, List, Single 구조로 분리해놓으신 이유가 있으실까요?

thxwelchs commented on src/main/java/dev/flab/studytogether/response/CommonResponse.java:
```diff
@@ -0,0 +1,22 @@
+
+import org.springframework.http.HttpStatus;
+
+public class CommonResponse {
```
공통 응답 객체를 일일이 모든 컨트롤러 메소드에 코딩하지 않고도, 공통적으로 적용할 수 있는 방법이 있을까요? 스프링에서는 이러한 핵심관심사와 공통적인 부가 관심사를 나누는 개념을 상당히 선호하고 있습니다. (AOP) 직접 구현하는 방법과, ControllerAdvice, RestControllerAdvice에 대해 학습해보세요!

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/member/controller/MemberAPIController.java:
```diff
@@ -1,22 +1,18 @@
 
-
-@RequestMapping("/members")
+@RequestMapping("/api/members")
```
/api prefix를 붙인 이유가 무엇일까요?

wooni97 replied:
API 엔드포인트와 다른 엔드포인트를 분리해서 일관성 있게 관리하려고 붙였습니다.

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/member/controller/MemberAPIController.java:
```diff
@@ -30,32 +26,39 @@ public MemberAPIController(MemberService memberService) {
-    public String join(MemberCreateRequestDto requestDto) {
-        memberService.createMember(requestDto.toServiceDto());
-        return "redirect:/login";
+    public SingleApiResponse<MemberResponse> join(MemberCreateRequestDto requestDto) {
+        Member member = memberService.createMember(requestDto.getId(), requestDto.getPassword(), requestDto.getNickname());
+        MemberResponse response = new MemberResponse(member.getSequenceId(), member.getId(), member.getNickname());
+
+        return SingleApiResponse.ok(response);
```
기존 구조에서 redirect를 하는 부분도 변경되었는데 이것도 의도된 변경일까요?

wooni97 replied:
네 그렇습니다. 우선은 api 설계와 구현에  집중하고 싶어 변경했습니다. 추후에 프론트단이 만들어지면 다시 생각해보려고요!

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/member/dto/MemberResponse.java:
```diff
@@ -0,0 +1,10 @@
+
+import lombok.Data;
+
+@Data
```
@Data 어노테이션은 어떤 역할을 하는것일까요? 이것을 사용함으로써 불필요한 어노테이션이 같이 사용되고 있진 않을까요?

wooni97 replied:
@toString + @getter + @setter + @RequiredArgsConstructor + @EqualsAndHashCode
위의 어노테이션을 한꺼번에 설정해주는 어노테이션인데, dto라서 추후에 해당 어노테이션들이 필요할것이라 생각했지만
지금으로 봤을땐 불필요한 어노테이션들이 더 많은게 맞습니다. 필요한 어노테이션만 사용하도록 수정할게요!

f-lab-inno replied:
좋습니다. @Data 어노테이션이 포함하고 있는 어노테이션들의 사용 목적을 명확하게 인지하고 사용해주시면 좋습니다!

f-lab-inno commented on src/main/java/dev/flab/studytogether/response/CommonResponse.java:
```diff
@@ -0,0 +1,22 @@
+
+import org.springframework.http.HttpStatus;
+
+public class CommonResponse {
```
공통 응답 객체를 일일이 모든 컨트롤러 메소드에 코딩하지 않고도, 공통적으로 적용할 수 있는 방법이 있을까요? 스프링에서는 이러한 핵심관심사와 공통적인 부가 관심사를 나누는 개념을 상당히 선호하고 있습니다. (AOP) 직접 구현하는 방법과, ControllerAdvice, RestControllerAdvice에 대해 학습해보세요!

wooni97 replied:
컨트롤러 메소드에 공통 응답 객체가 명시돼있는게 코드를 보는 입장에서 편하다고 생각하여 그렇게 구현했는데
고민을 해보니 반복적인 작업이 맞고, 혹여나 나중에 응답 객체와 관련하여 변경이 일어난다면 api 갯수만큼 변경해야하는 경우가 생길것 같아 공통 부가 관심사로 분리하는게 맞다고 생각이 드네요! 이 부분은 더 학습해보고 수정하여 새롭게 pr 올리겠습니다.

f-lab-inno commented on src/main/java/dev/flab/studytogether/response/ListApiResponse.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.List;
+
+public class ListApiResponse<T> extends CommonResponse {
```
SingleApiResponse와 비교해보았을 때, message 필드 보유 여부 외에는 큰 차이가 없어보이는데, List, Single 구조로 분리해놓으신 이유가 있으실까요?

wooni97 replied:
컨트롤러에서 메서드를 보고 해당 API가 단일 객체를 다루는지, 여러 객체를 다루는지 명확하게 전달하고 싶어서 분리한게 가장 큰 이유입니다.

wooni97 replied:
어차피 제네릭을 사용하고 있고, 저번에 말씀 주신대로 응답 data에 List 외에도 다른 자료구조가 들어갈 수 있는 확장성을 고려하여 List, Single로 분리하는게 불필요해보여서 수정했습니다!

wooni97 committed:
refactor: 공통 API 응답 객체 수정

기존의 CommonResponse를 확장한 SingleApiResponse, ListApiResponse를 삭제하고 ApiResponse로 통합

wooni97 committed:
refactor: 공통 API 응답 객체 수정에 따른 메서드 반환 타입 수정

wooni97 committed:
refactor: 공통 API 응답 객체 수정에 따른 메서드 반환 타입 수정

wooni97 committed:
refactor: 어노테이션 수정

@Data 어노테이션이 불필요한 어노테이션을 포함하고 있어, 필요한 어노테이션들로만 수정
